File: cpp/lexer.cpp
#include "lexer.h"
#include "token.h"
#include <stdio.h>
#include <ctype.h>
#include <fstream>
#include <unordered_set> 

std::unordered_set<std::string> keywords        ( {"var", "if", "then", "else", "for", "while", "do", "in", "function", "return"} );
std::unordered_set<std::string> symbols         ( {"{", "}", "(", ")", "[", "]", "<", ">", "+", "-", "*", "/", "%", "!", "?", "=", ".", ",", "&", "|", ";", ":", "$"  } );
std::unordered_set<std::string> multi           ( { "==", "!=", ">=", "<=", "+=", "-=", "*=", "/=", "%=", "<[", "]>" } );

bool is_id_char(char c) {
    return isdigit(c) || isalpha(c) || c == '_';
}

std::vector<Token> lex_string ( std::string &source ) {

    std::vector<Token> tokens;
    int index = 0;

    while ( index < source.size() ) {
        
        if ( isspace(source[index]) ) {
            // could treat newlines special, but don't feel like it yet
            while( index < source.size() && isspace(source[index]) ) {
                ++index;
            }
        }
        else {
            // identifiers and keywords (reserved words)
            if ( isalpha(source[index]) || source[index] == '_' ) {
                int end = index + 1;

                // move ahead until the end of the identifier
                while ( end < source.size() && is_id_char(source[end]) ) {
                    ++end;
                }
                std::string word = source.substr(index, end - index);
                Token token;

                // unordered_set.contains() only came out in c++20, so count > 0 is a less-readable way to check set membership
                if ( keywords.count(word) > 0 ) {
                    token = Token { word, keyword };
                }
                else if (word == "true" || word == "false") {
                    token = Token { word, bool_literal };
                }
                else {
                    token = Token { word, identifier };
                }
                tokens.push_back(token);
                index = end;
            }
            // integers and real numbers (floating point)
            else if ( isdigit(source[index]) ) {
                int end = index + 1;

                while ( end < source.size() && isdigit(source[end]) ) {
                    ++end;
                }
                Token token;

                if ( source[end] == '.' ) {
                    ++end;
                    while ( end < source.size() && isdigit(source[end]) ) {
                        ++end;
                    }
                    std::string real = source.substr(index, end - index);
                    token = Token { real, real_literal };
                }
                else {
                    std::string integer = source.substr(index, end - index);
                    token = Token { integer, int_literal };
                }
                tokens.push_back(token);
                index = end;
            }
            // string literals
            else if ( source[index] == '\"' || source[index] == '\'' ) {
                char quote = source[index];
                int end = index + 1;
                while ( end < source.size() && source[end] != quote ) {
                    ++end;
                }
                std::string string_value = source.substr(index + 1, end - (index + 1));
                Token token = Token { string_value, string_literal };
                tokens.push_back(token);
                index = end + 1;
            }
            // symbols and operators
            else if (symbols.count(std::string(1, source[index])) > 0) {
                int end = index + 2;
                Token token;
                if ( end <= source.size() && ( multi.count(source.substr(index, end - index)) > 0) ) {
                    std::string symbol_value = source.substr(index, end - index);
                    token = Token { symbol_value, symbol };
                    index = end;
                }
                else {
                    token =  Token { std::string(1, source[index]), symbol };
                    ++index;
                }
                tokens.push_back(token);
            }
            else {
                printf("Invalid start of token %c\n", source[index]);
                break;
            }

        }
    }
    return tokens;
}

void pretty_print( std::vector<Token> &tokens ) {
    printf("[\n");
    for (int i = 0; i < tokens.size(); ++i) {
        const char* token = tokens[i].sval.c_str();
        const char* token_type = tokenTypeString( tokens[i].type).c_str();
        printf("%s : %s\n", token, token_type);
    }
    printf("]\n");
}

void tuple_print( std::vector<Token> &tokens ) {
    printf("[\n");

    int i = 0;
    while (i < tokens.size() - 1) {
        printf("(\"%s\", \"%s\"),\n", tokens[i].sval.c_str(), tokenTypeString( tokens[i].type).c_str());
        ++i;
    }
    printf("(\"%s\", \"%s\")\n", tokens[i].sval.c_str(), tokenTypeString( tokens[i].type).c_str());

    printf("]\n");
}


int main( int argc, const char* argv[] ) {
    // for now just take an argument string and tokenize it, worry about file IO later
    if( argc > 1 ) {
        std::string source = argv[1];

        // if file flag is used, read from file instead 
        if( source == "-f" && argc > 2 ) {
            std::string source_file = argv[2];
            std::ifstream ifs(source_file);
            source.assign( (std::istreambuf_iterator<char>(ifs) ), (std::istreambuf_iterator<char>() ) );
        } 


        std::vector<Token> tokens = lex_string(source);

        tuple_print(tokens);
    }
    return 0;
}

File: Python/lexer.py
import sys


# lexer helper methods

keywords = {"var", "if", "then", "else", "for", "while", "do", "in", "function", "return"}


symbols = {
    "{" : "left-brace",
    "}" : "right-brace",
    "(" : "left-paren",
    ")" : "right-paren",
    "[" : "left-bracket",
    "]" : "right-bracket",
    "<" : "less-than",
    ">" : "greater-than",
    "+" : "plus",
    "-" : "minus",
    "*" : "star",
    "/" : "slash",
    "%" : "percent",
    "!" : "not",
    "?" : "question",
    "=" : "equals",
    "." : "dot",
    "," : "comma",
    "&" : "and", 
    "|" : "or",
    ";" : "semicolon",
    ":" : "colon"
}

maybe_multichar = {
    "=", "!", ">", "<", "+", "-", "*", "/", "%"
}

multi = {
    "==" : "equals-equals",
    "!=" : "not-equals",
    ">=" : "greater-equals",
    "<=" : "less-equals",
    "+=" : "plus-equals",
    "-=" : "minus-equals",
    "*=" : "star-equals",
    "/=" : "slash-equals",
    "%=" : "percent-equals"
}

def pretty_print(ts):
    print("[")
    for t in ts:
        print(f"{t[0]} : {t[1]}")
    print("]")

def tuple_print( tokens ):
    print("[")

    i = 0
    while i < (len(tokens) - 1):
        print("(\"%s\", \"%s\")," % (tokens[i][1], tokens[i][0]))
        i += 1
    print("(\"%s\", \"%s\")" % (tokens[i][1], tokens[i][0]))

    print("]")


def is_id_char(c):
    """returns true if the passed character is a valid identifier character"""
    return c.isalnum() or (c == "_")


def lex_string(source, print_tokens=False):
    global keywords
    global symbols
    
    tokens = []
    i = 0

    while(i < len(source)):

        if source[i].isspace():
            while i < len(source) and source[i].isspace():
                i += 1
        else:
            # identifiers
            if source[i].isalpha() or source[i] == "_":
                j = i + 1
                while j < len(source) and is_id_char(source[j]):
                    j += 1

                word = source[i : j]
                
                if word in keywords:
                    tokens.append( ('keyword', word) )
                elif word in ('true', 'false'):
                    tokens.append( ('bool-literal', word))
                else:
                    tokens.append( ('identifier', source[i : j]) )
                i = j
            
            elif source[i].isnumeric():
                j = i + 1
                while j < len(source) and source[j].isnumeric():
                    j += 1
                if source[j] == ".":
                    j += 1
                    while j < len(source) and source[j].isnumeric():
                        j += 1
                    tokens.append( ('real-literal', source[i : j]) )
                    i = j
                else:
                    tokens.append( ('int-literal', source[i : j]) )
                    i = j

            elif source[i] in "\'\"":
                quote = source[i]
                j = i + 1
                while source[j] != quote:
                    j += 1
                tokens.append( ('string-literal', source[ i + 1 : j ]) )
                i = j + 1

            elif source[i] in symbols.keys():
                
                j = i + 2
                if j <= len(source) and source[i : j] in multi.keys():
                    symbol = source[i : j]
                    #tokens.append( (multi[symbol], symbol) )
                    tokens.append( ('symbol', symbol) )
                    i = j
                else:
                    symbol = source[i : i+1]
                    #tokens.append( (symbols[symbol], symbol) )
                    tokens.append( ('symbol', symbol) )
                    i += 1
            else:
                raise Exception("Invalid token start: " + source[i])
    
    #pretty_print(tokens)
    if print_tokens:
        tuple_print(tokens)

    # swap tuple order so that it works with the lexer and matches the order of the c++ lexer
    return list( map(lambda t : (t[1], t[0]), tokens) )
                
                 
def lex_file(path):
    with open (path, 'r') as sourcefile:
        return lex_string(''.join(sourcefile.readlines()), print_tokens=False)


if __name__ == "__main__":

    opts = [opt for opt in sys.argv[1:] if opt.startswith("-")]
    args = [arg for arg in sys.argv[1:] if not arg.startswith("-")]

    
    if "-f" in opts:
        if len(args) == 0:
            raise SystemExit('Expected path to file')
        else:
            path_to_sourcefile = args[0]
            with open (path_to_sourcefile, 'r') as sourcefile:
                lex_string(''.join(sourcefile.readlines()), print_tokens=True)
    else:
        if len(args) == 0:
            raise SystemExit('Command mode requires at least one argument')
        else:
            lex_string(' '.join(args), print_tokens=True)